src/xopen/__init__.py

```diff
@@ -551,6 +558,39 @@ def _open_zst(
     return io.BufferedWriter(f)  # mode "ab" and "wb"
 
 
+def _open_lz4(
+    filename: FileOrPath,
+    mode: str,
+    compresslevel: Optional[int],
+    threads: Optional[int],
+):
+    assert mode in ("rb", "ab", "wb")
+    if compresslevel is None:
+        compresslevel = XOPEN_DEFAULT_LZ4_COMPRESSION
+
+    if threads != 0:
+        try:
+            return _PipedCompressionProgram(
+                filename,
+                mode,
+                compresslevel,
+                threads,
+                program_settings=_PROGRAM_SETTINGS["lz4"],
+            )
+        except OSError:
+            if lz4 is None:
+                # No fallback available
+                raise
+
+    if lz4 is None:
+        raise ImportError("lz4 module not available")
```

Comment:
Looks good. What do we do when threads==0 but the lz4 program is available? Now it crashes, my inclination is to keep it that way if that is consistent with the behavior for the other formats. My idea is to always do what the user explicitly requested and crash if that is not possible. Although a case could be made that xopen should only take that as a suggestion, and prioritize getting the file open.

I just realized that this becomes irrelevant when lz4 is not an optional dependency anymore. Then this code can be simplified as well.

Reply:
Yes. So in that case, lz4 CLI support can be slashed and the lz4 library be made a mandatory dependency. That should make the code much simpler. Supporting CLI programs was a mainstay of xopen back when there where no extra fast gzip libraries available and the fastest way to get things done was through a CLI program. 

But please benchmark first. Getting threaded performance with GIL dropping faster than a C CLI application is non-trivial. GIL-drop GIL-retain cycles have overhead and this can lead to behavior where the cores are not used to their full potential.